### PR TITLE
Added support for EUF infix notation

### LINKED ISSUE #239: Overload infix notation for  EUF
The only theory that does not currently support of infix notation is UF. This issue is to discuss possible problems and syntax issues related to extending UF with infix notation. In particular, a possible idea is to use a syntax like

```
f(3) #if f is a function symbol, this should return the application of 3 to f
```

It is unclear how much interest there is in this. Also, a few examples would be useful to better understand the implications of this.

### PR DESCRIPTION
**Pull request by mikand**
This closes #239

### CODE REVIEW

File: pysmt/fnode.py [hunk 1]
```diff
@@ -854,6 +861,11 @@ def __rshift__(self, right):
 
     def __mod__(self, right):
         return self._apply_infix(right, None, bv_function=_mgr().BVURem)
+
+    def __call__(self, *args):
```
Comment (marcogario) [inline pysmt/fnode.py]: If args is empty ``f()`` the iterator will not execute and neither  ``_infix_prepare_args``. 

Also, don't you should not check against the type of the function, but the type of the argument. See comment below.

File: pysmt/test/test_euf.py [hunk 2]
```diff
@@ -37,6 +37,22 @@ def test_euf(self):
...
+        f = Symbol("f", ftype1)
+        g = Symbol("g", ftype2)
+
+        with self.assertRaises(PysmtModeError):
+            f(1.0)
+
+        get_env().enable_infix_notation = True
+        check = Equals(f(1.0), g(2.0, 4))
+
+        self.assertSat(check, logic=UFLIRA,
```
Comment (marcogario) [inline pysmt/test/test_euf.py]: We do not need a solver for this. Let's just check that the infix and prefix formula yield the same result.

File: pysmt/test/test_euf.py [hunk 3]
```diff
@@ -37,6 +37,22 @@ def test_euf(self):
         self.assertSat(check, logic=UFLIRA,
                        msg="Formula was expected to be sat")
 
+    @skipIfNoSolverForLogic(UFLIRA)
+    def test_infix(self):
+        ftype1 = FunctionType(REAL, [REAL])
+        ftype2 = FunctionType(REAL, [REAL, INT])
```
Comment (marcogario) [inline pysmt/test/test_euf.py]: Can you add a test case for f: Int x BV8 -> Real? It seems to me that the implementation that you provide should not work with BV. When you convert the args, you should look into the function type definition and for each positional argument detect the type... I am confused on why this works for INT...

File: pysmt/fnode.py (hunk 1 above)
Comment (mikand) [reply pysmt/fnode.py]: Done.

File: pysmt/test/test_euf.py (hunk 2 above)
Comment (mikand) [reply pysmt/test/test_euf.py]: OK

File: pysmt/test/test_euf.py (hunk 3 above)
Comment (mikand) [reply pysmt/test/test_euf.py]: Done.

**Pull request merged**